File: FileThumbnails/js/js/scenario3.js
//// THIS CODE AND INFORMATION IS PROVIDED "AS IS" WITHOUT WARRANTY OF
//// ANY KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO
//// THE IMPLIED WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A
//// PARTICULAR PURPOSE.
////
//// Copyright (c) Microsoft Corporation. All rights reserved

(function () {
    "use strict";

    var page = WinJS.UI.Pages.define("/html/scenario3.html", {
        ready: function (element, options) {
            document.getElementById("document-thumb-start").addEventListener("click", pickDocument, false);
        }
    });

    function pickDocument() {
        // Set API parameters
        var requestedSize = 100,
            thumbnailMode = Windows.Storage.FileProperties.ThumbnailMode.documentsView;

        // Clean output in case of repeat usage
        cleanOutput();

        var openpicker = new Windows.Storage.Pickers.FileOpenPicker();
        openpicker.fileTypeFilter.replaceAll([".doc", ".xls", ".ppt", ".docx", ".xlsx", ".pptx", ".pdf", ".txt", ".rtf"]);
        openpicker.suggestedStartLocation = Windows.Storage.Pickers.PickerLocationId.documentsLibrary;
        openpicker.pickSingleFileAsync().done(function (file) {
            if (file) {
                file.getThumbnailAsync(thumbnailMode, requestedSize).done(
                    function (thumbnail) {
                        if (thumbnail) {
                            outputResult(file, thumbnail, "ThumbnailMode.documentsView", requestedSize);
                        } else {
                            WinJS.log && WinJS.log(SdkSample.errors.noIcon, "sample", "status");
                        }
                    },
                    function (error) {
                        WinJS.log && WinJS.log(SdkSample.errors.fail, "sample", "status");
                    }
                );
            } else {
                WinJS.log && WinJS.log(SdkSample.errors.cancel, "sample", "status");
            }
        });
    }

    function outputResult(item, thumbnailImage, thumbnailMode, requestedSize) {
        document.getElementById("document-thumb-imageHolder").src = URL.createObjectURL(thumbnailImage, { oneTimeOnly: true });
        // Close the thumbnail stream once the image is loaded
        document.getElementById("document-thumb-imageHolder").onload = function () {
            thumbnailImage.close();
        };
        document.getElementById("document-thumb-modeName").innerText = thumbnailMode;
        document.getElementById("document-thumb-fileName").innerText = "File used: " + item.name;
        document.getElementById("document-thumb-requestedSize").innerText = "Requested size: " + requestedSize;
        document.getElementById("document-thumb-returnedSize").innerText = "Returned size: " + thumbnailImage.originalWidth + "x" + thumbnailImage.originalHeight;
    }

    function cleanOutput() {
        WinJS.log && WinJS.log("", "sample", "status");
        document.getElementById("document-thumb-imageHolder").src = "/images/placeholder-sdk.png";
        document.getElementById("document-thumb-modeName").innerText = "";
        document.getElementById("document-thumb-fileName").innerText = "";
        document.getElementById("document-thumb-requestedSize").innerText = "";
        document.getElementById("document-thumb-returnedSize").innerText = "";
    }
})();

File: Samples/FileThumbnails/cppwinrt/Scenario3.cpp
//*********************************************************
//
// Copyright (c) Microsoft. All rights reserved.
// This code is licensed under the MIT License (MIT).
// THIS CODE IS PROVIDED *AS IS* WITHOUT WARRANTY OF
// ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING ANY
// IMPLIED WARRANTIES OF FITNESS FOR A PARTICULAR
// PURPOSE, MERCHANTABILITY, OR NON-INFRINGEMENT.
//
//*********************************************************

#include "pch.h"
#include "Scenario3.h"
#include "Scenario3.g.cpp"
#include "SampleConfiguration.h"

using namespace winrt;
using namespace Windows::Foundation;
using namespace Windows::Storage;
using namespace Windows::Storage::FileProperties;
using namespace Windows::Storage::Pickers;
using namespace Windows::UI::Xaml;
using namespace Windows::UI::Xaml::Navigation;

namespace winrt::SDKTemplate::implementation
{
    Scenario3::Scenario3()
    {
        InitializeComponent();
    }

    void Scenario3::OnNavigatedTo(NavigationEventArgs const&)
    {
        Helpers::ResetOutput(rootPage, ThumbnailImage(), OutputTextBlock());
    }

    fire_and_forget Scenario3::GetThumbnailButton_Click(IInspectable const&, RoutedEventArgs const&)
    {
        auto lifetime = get_strong();

        Helpers::ResetOutput(rootPage, ThumbnailImage(), OutputTextBlock());

        // Pick a document
        FileOpenPicker openPicker;
        openPicker.FileTypeFilter().ReplaceAll(FileExtensions::Document());
        StorageFile file = co_await openPicker.PickSingleFileAsync();

        if (file != nullptr)
        {
            const ThumbnailMode thumbnailMode = ThumbnailMode::DocumentsView;
            const uint32_t size = 100;
            StorageItemThumbnail thumbnail = co_await file.GetThumbnailAsync(thumbnailMode, size);
            if (thumbnail != nullptr)
            {
                Helpers::DisplayResult(ThumbnailImage(), OutputTextBlock(), thumbnailMode, size, file, thumbnail, false);
                thumbnail.Close();
            }
            else
            {
                rootPage.NotifyUser(Errors::NoIcon, NotifyType::StatusMessage);
            }
        }
        else
        {
            rootPage.NotifyUser(Errors::Cancel, NotifyType::StatusMessage);
        }
    }
}
